pkg/pubsub/system.go

```diff
@@ -36,24 +33,27 @@ func (s *System) Publish(ctx context.Context, connection string, topic string, m
 
 func (s *System) UpsertConnection(ctx context.Context, config interface{}, name string, provider string) error {
 	s.mux.Lock()
-	defer s.mux.Unlock()
 	// Check if the connection already exists.
 	if conn, ok := s.connections[name]; ok {
 		// If the provider is the same, update the existing connection.
 		if s.providers[name] == provider {
 			return conn.UpdateConnection(ctx, config)
 		}
-		// Otherwise, close the existing connection and create a new one.
-		if err := s.CloseConnection(name); err != nil {
-			return err
-		}
 	}
+	s.mux.Unlock()
```

Comment:
There is a possibility of returning without unlocking the mutex (`return conn.UpdateConnection`), which would lead to a deadlock on subsequent calls.

Why not hold the mutex for the full function call? Does `s.CloseConnection()` also acquire the mutex? Can you wrap the CloseConnection() internals in a private, non-blocking function that can be shared between the public CloseConnection() and UpsertConnection()?